bson/__init__.py

```diff
@@ -1006,7 +1006,10 @@ def _dict_to_bson(
             elements.append(_name_value_to_bson(b"_id\x00", doc["_id"], check_keys, opts))
         for key, value in doc.items():
             if not top_level or key != "_id":
-                elements.append(_element_to_bson(key, value, check_keys, opts))
+                try:
+                    elements.append(_element_to_bson(key, value, check_keys, opts))
+                except InvalidDocument as err:
+                    raise InvalidDocument(f"Invalid document {doc} | {err}") from err
```

Comment:
I'm not sure this is the pattern we want, in the C version of this code we recursively call write_dict() for subdocuments which means if we end up erroring in a nested field the error will be something like:
`Invalid document {"a": {"b": {"c": ...}}} | Invalid document {"b": {"c": ...}} | Invalid document {"c": ...} | ...`

Is that the intended behavior?